src/db/mod.rs

```diff
@@ -353,15 +355,61 @@ impl Database {
 
     /// Runs a database-level command using the provided `ClientSession`.
     ///
-    /// Note that no inspection is done on `doc`, so the command will not use the database's default
-    /// read concern or write concern. If specific read concern or write concern is desired, it must
-    /// be specified manually.
+    /// If the `ClientSession` provided is currently in a transaction, `command` must not specify a
+    /// read concern. If this operation is the first operation in the transaction, the read concern
+    /// associated with the transaction will be inherited.
+    ///
+    /// Otherwise no inspection is done on `command`, so the command will not use the database's
+    /// default read concern or write concern. If specific read concern or write concern is
+    /// desired, it must be specified manually.
     pub async fn run_command_with_session(
         &self,
-        command: Document,
+        mut command: Document,
         selection_criteria: impl Into<Option<SelectionCriteria>>,
         session: &mut ClientSession,
     ) -> Result<Document> {
+        let mut selection_criteria = selection_criteria.into();
```

Comment:
I think this logic is more suited to `RunCommand::build`. You'll need to pass the transaction state into `RunCommand::new` but I think that's okay.

Reply:
We also added the logic for adding the transaction's read concern to the first command in a transaction to `execute_operation` (see `executor.rs` line 277). This was done for commands that don't accept a read concern via their options, but it also applies here. Now that I think about it, it also applies to the commands which _do_ take a read concern, meaning we can also remove the logic for adding it to the options structs in the `resolve_read_concern_with_session!` macro too. Then the read concern and write concern macros will look pretty much the same. 

Getting back to `run_command`, the validation that the command document doesn't contain a read concern does need to remain here though, just like we still need to verify the options structs don't contain read concerns / write concerns.

Reply:
Oh gotcha, yeah that was just for the write commands but seems reasonable to move all of the logic there instead.